pg_search/src/query/mod.rs

```diff
@@ -760,127 +764,234 @@ impl SearchQueryInput {
+                    subqueries.push((Occur::Should, query));
+                    should_children.push(output);
                 }
+
+                let mut must_not_children = vec![];
                 for input in must_not {
-                    subqueries.push((
-                        Occur::MustNot,
-                        input.into_tantivy_query(
-                            schema,
-                            parser,
-                            searcher,
-                            index_oid,
-                            relation_oid,
-                            expr_context,
-                            planstate,
-                        )?,
-                    ));
+                    let output = recurse(input)?;
+                    let query = B::clone_query(&output);
+                    subqueries.push((Occur::MustNot, query));
+                    must_not_children.push(output);
+                }
+
+                let query = Box::new(BooleanQuery::new(subqueries));
+
+                // Create children in order: must, should, must_not
+                let mut children = Vec::new();
+                for (idx, child) in must_children.into_iter().enumerate() {
+                    let child_query = B::extract_query(&child);
+                    let wrapped = builder.build_with_children(
+                        &SearchQueryInput::Empty,
+                        child_query.box_clone(),
+                        || format!("Must Clause [{}]", idx),
+                        || vec![child],
+                    );
+                    children.push(wrapped);
+                }
+                for (idx, child) in should_children.into_iter().enumerate() {
+                    let child_query = B::extract_query(&child);
+                    let wrapped = builder.build_with_children(
+                        &SearchQueryInput::Empty,
+                        child_query.box_clone(),
+                        || format!("Should Clause [{}]", idx),
+                        || vec![child],
+                    );
+                    children.push(wrapped);
                 }
-                Ok(Box::new(BooleanQuery::new(subqueries)))
+                for (idx, child) in must_not_children.into_iter().enumerate() {
+                    let child_query = B::extract_query(&child);
+                    let wrapped = builder.build_with_children(
+                        &SearchQueryInput::Empty,
+                        child_query.box_clone(),
+                        || format!("MustNot Clause [{}]", idx),
+                        || vec![child],
+                    );
+                    children.push(wrapped);
+                }
+
+                Ok(builder.build_with_children(
+                    &SearchQueryInput::Boolean {
+                        must: Vec::new(),
+                        should: Vec::new(),
+                        must_not: Vec::new(),
+                    },
```

Comment:
Sorry: passing as a reference doesn't help if we're still making a brand new struct to pass... I suppose that something like:
```suggestion
                    &self,
```

... does not work here, because it will claim that "`self` has partially moved" or something, do to having consumed `must`/`should`/`not`?